src/core/prerender/utils.ts

```diff
@@ -34,12 +33,17 @@ export function extractLinks(
 
   // Extract from any <TAG href=""> to crawl
   if (crawlLinks) {
-    _links.push(
-      ...[...html.matchAll(LINK_REGEX)]
-        .map((m) => escapeHtml(m[1]))
-        .filter((m) => !decodeURIComponent(m).startsWith("#"))
-        .filter((link) => allowedExtensions.has(getExtension(link)))
-    );
+    walkSync(parseHTML(html), (node) => {
+      if (node.attributes.href) {
```

Comment:
Can you please refactor this to do with early return?

Also we can have a new crawler option that gets node and returns false if node can be ignored wdyt?

Reply:
oh, I like the idea, but it would tie us to this particular AST. maybe better seeing if we need it? Or maybe adding a hook that gets called with extracted links instead?

Reply:
my thinking was to give full control/awareness to filter callback (node type, other atts combo, etc). personally, I'm fine with betting on ultrahtml AST. But also fine with first adding that generic hook, feel free to drop another PR 👍🏼